cli/NetworkIO.hpp

```diff
@@ -0,0 +1,286 @@
+    DCHECK(response_ready_);
+    return canceled_;
+  }
+
+ private:
+  bool response_ready_;
+  bool canceled_;
+  QueryRequest const * request_;
+  QueryResponse response_message_;
+  Mutex mutex_;
+  ConditionVariable *condition_;  // note: owned by the mutex.
+
+  DISALLOW_COPY_AND_ASSIGN(RequestState);
+};
+
+}  // namespace networkio_internal
+
+using networkio_internal::RequestState;
+
+/**
+ * @brief Contains the callback methods which the gRPC service defines.
+ * When a request is made of gRPC, a gRPC worker thread is spun up and enters one of the callback methods
+ * (eg SendQuery). This thread keeps the network connection open while Quickstep processes the query. Concurrency
+ * control between the gRPC worker thread, and the Quickstep thread is controlled by a RequestState object which is
+ * created for each interaction.
+ */
+class NetworkCliServiceImpl final : public NetworkCli::Service {
+ public:
+  NetworkCliServiceImpl()
+      : running_(true) {}
+
+  /**
+   * @brief Handles gRPC request.
+   * Sets the buffer in the RequestState, places the request on a queue, then waits for a response. The response shall
+   * be triggered by a
+   */
+  Status SendQuery(grpc::ServerContext *context,
+                   const QueryRequest *request,
+                   QueryResponse *response) override {
+    std::unique_ptr<RequestState> request_state;
+    // Check to see if the gRPC service has been shutdown.
+    {
+      std::unique_lock<Mutex> lock(service_mtx_);
+      if (!running_) {
+        return Status::CANCELLED;
+      }
+      // While we have a service lock, we add to the Queue. Note that we keep the service lock to protect ourselves from
+      // a race condition in the kill() method.
+      request_state.reset(new RequestState(request));
+
+      // Pushing to the queue will notify consumers.
+      request_queue_.push(request_state.get());
+    }
+    DCHECK(request_state);
+
+    // We have pushed to the request queue, so all there is to do now is wait for Quickstep to process the request.
+    request_state->waitForResponse();
+    if (request_state->getCanceled()) {
+      return Status::CANCELLED;
+    }
+    *response = request_state->getResponse();
+    return Status::OK;
+  }
+
+  /**
+   * @brief The consumer thread waits for a request to materialize.
+   * @return A non-owned RequestState.
+   */
+  RequestState* waitForRequest() {
+    return request_queue_.popOne();
+  }
+
+  /**
+   * @brief Stops accepting further requests and cancels all pending requests.
+   */
+  void kill() {
+    {
+      // This action guarantees that no further requests are added to the queue.
+      std::unique_lock<Mutex> lock(service_mtx_);
+      running_ = false;
+    }
+    // Go through each pending request, and cancel them.
+    while (!request_queue_.empty()) {
+      request_queue_.popOne()->cancel();
+    }
+  }
+
+ private:
+  Mutex service_mtx_;
+  bool running_;
+  ThreadSafeQueue<RequestState*> request_queue_;
```

Comment:
I suggest to follow [the official async server example](https://github.com/grpc/grpc/blob/master/examples/cpp/helloworld/greeter_async_server.cc#L54) so that we don't need mutexes and locks for multiple concurrent requests.